lib/mix/gear_static_analysis.ex

```diff
@@ -152,68 +152,62 @@ defmodule Mix.Tasks.Compile.GearStaticAnalysis do
     end
   end
 
-  defp check_use_within_module(mod, _kw, _meta, file, _tool?) do
-    cond do
-      mod == Gettext ->
-        {:error, file, [],
-         "directly invoking `use Gettext` is not allowed (`use Antikythera.Gettext` instead)"}
-
-      true ->
-        nil
-    end
+  defp check_use_within_module(Gettext, _kw, _meta, file, _tool?) do
+    {:error, file, [],
+     "directly invoking `use Gettext` is not allowed (`use Antikythera.Gettext` instead)"}
   end
 
-  defp check_remote_call(mod, fun, args, meta, file, tool?) do
-    use_internals? = use_antikythera_internal_modules?()
+  defp check_use_within_module(_mod, _kw, _meta, _file, _tool?), do: nil
 
-    if !tool? do
-      cond do
-        mod == System and fun in [:halt, :stop] ->
-          {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  defguardp writing_to_stdout?(args)
+            when is_list(args) and (length(args) == 1 or hd(args) == :stdio)
 
-        mod == :erlang and fun == :halt ->
-          {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  defp check_remote_call(System, fun, _args, meta, file, false) when fun in [:halt, :stop] do
+    {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  end
 
-        mod == :init ->
-          {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  defp check_remote_call(:erlang, :halt, _args, meta, file, false) do
+    {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  end
 
-        mod == IO and fun in [:inspect, :puts, :write] and writing_to_stdout?(args) ->
-          severity = if Mix.env() == :prod, do: :error, else: :warning
+  defp check_remote_call(:init, _fun, _args, meta, file, false) do
+    {:error, file, meta, "disturbing execution of ErlangVM is strictly prohibited"}
+  end
 
-          {severity, file, meta,
-           "writing to STDOUT/STDERR is not allowed in prod environment (use each gear's logger instead)"}
+  defp check_remote_call(IO, fun, args, meta, file, false)
+       when fun in [:inspect, :puts, :write] and writing_to_stdout?(args) do
+    severity = if Mix.env() == :prod, do: :error, else: :warning
 
-        mod in [Process, Task, Agent] and spawning_a_new_process?(Atom.to_string(fun)) ->
-          {:error, file, meta, "spawning processes in gear's code is prohibited"}
+    {severity, file, meta,
+     "writing to STDOUT/STDERR is not allowed in prod environment (use each gear's logger instead)"}
+  end
 
-        mod == :os and fun == :cmd ->
-          {:error, file, meta, "calling :os.cmd/1 in gear's code is prohibited"}
+  defp check_remote_call(mod, fun, _args, meta, file, false) when mod in [Process, Task, Agent] do
+    if spawning_a_new_process?(Atom.to_string(fun)) do
```

Comment:
(optional)
Because this suggestion largely a matter of taste, you don't have to accept it.

The current (new) implementation slightly change the original logic; i.e., when `spawning_a_new_process?(Atom.to_string(fun))` is false, the current implementation doesn't check following conditions.
For now, this doesn't mean to change `check_remote_call/6`'s behavior.
But I think this might become a constraint when introducing new check conditions.

In addition, I think just using a multi-clauses function doesn't really solve the function's complexity.

How about breaking `check_remote_call/6` into smaller functions?
The below is my incomplete and untested suggestion:

```elixir
defp check_remote_call(mod, fun, args, meta, file, false) do
  with :ok <- check_disturbing_execution_of_vm(mod, fun),
       :ok <- check_writing_to_stdout(mod, fun, args) do
       # add more checking functions here
    nil
  else
    {severity, message} -> {severity, file, meta, message}
  end
end

defp check_remote_call(_mod, _fun, _args, _meta, _file, _tool?), do: nil

defp check_disturbing_execution_of_vm(mod, fun) do
  if disturbing_execution_of_vm?(mod, fun) do
    {:error, "disturbing execution of ErlangVM is strictly prohibited"}
  else
    :ok
  end
end

defp disturbing_execution_of_vm?(System, fun) when fun in [:halt, :stop], do: true
defp disturbing_execution_of_vm?(:erlang, :halt), do: true
defp disturbing_execution_of_vm?(:init, _fun), do: true
defp disturbing_execution_of_vm?(_mod, _fun), do: false

defp check_writing_to_stdout(IO, fun, args) when fun in [:inspect, :puts, :write] do
  if writing_to_stdout?(args) do
    severity = if Mix.env() == :prod, do: :error, else: :warning

    {severity,
      "writing to STDOUT/STDERR is not allowed in prod environment (use each gear's logger instead)"}
  else
    :ok
  end
end

defp check_writing_to_stdout(_mod, _fun, _args), do: :ok
```

Reply:
Applied the suggestion at 7ad3f41.